health/service.go

```diff
@@ -0,0 +1,151 @@
+package health
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"sync"
+	"time"
+)
+
+// CheckPath is the baked in SOH endpoint path.
+const CheckPath = "/soh"
+
+// Service provides a mechanism to update a service SOH status.
+type Service struct {
+	mu sync.Mutex
+
+	// status is used to indicate whether the service is running
+	status bool
+	// last stores the time of the last update.
+	last time.Time
+
+	// start stores when the service was started.
+	start time.Time
+	// aged is the time if no updates have happened indicates the service is no longer running.
+	// set to 0 if no age check needed
+	aged time.Duration
+	// startup is the time after the start which the check is assumed to be successful.
+	startup time.Duration
+}
+
+// New returns a health Service which provides running SOH capabilities.
+func New(endpoint string, aged, startup time.Duration) *Service {
+	service := &Service{
+		aged:    aged,
+		last:    time.Now(),
+		start:   time.Now(),
+		startup: startup,
+	}
+
+	router := http.NewServeMux()
+	router.HandleFunc(CheckPath, service.handler)
+
+	srv := &http.Server{
+		Addr:              endpoint,
+		Handler:           router,
+		ReadHeaderTimeout: 2 * time.Second,
+	}
+
+	go func() {
+		_ = srv.ListenAndServe()
+	}()
+
+	return service
+}
+
+// state returns the current application state, this is likely to
+// be expanded as new checks are added.
+func (s *Service) state() bool {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	return s.status
+}
+
+func (s *Service) handler(w http.ResponseWriter, r *http.Request) {
+	switch ok := s.state(); {
+	case time.Since(s.start) < s.startup:
+		// the check has been made too soon, this is to avoid
+		// a service being terminated before the initial check
+		w.WriteHeader(http.StatusOK)
```

Comment:
I think UNKOWN is shown when there is no health check, it is reasonable to return OK here as the application is running, otherwise it will cause a restart. But this is Mark's code, i haven't discussed the logic with him